File: Permutations.cpp
class Solution {
public:

    void solve(vector<int>&nums, vector<vector<int>>& res, int i ){

        if(i>= nums.size()){
            res.push_back(nums);
            return ;
        }

        for(int idx =i;idx<nums.size();idx++){
            swap(nums[i], nums[idx]);
            solve(nums, res, i+1);
            swap(nums[i], nums[idx]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {

        vector<vector<int>>res;
        solve(nums, res,0);
        return res;
        
    }
};

File: Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:

        def solve(i, nums, res):

            if i>= len(nums):
                res.append(nums[:])
                return

            for idx in range(i,len(nums)):

                nums[i], nums[idx] = nums[idx], nums[i]
                solve(i+1, nums,res)
                nums[i],nums[idx]= nums[idx],nums[i]

        res = []
        solve(0,nums, res)
        return res   